### PR TITLE
port: add ChildIP

### PR DESCRIPTION
**Pull request by giuseppe**
allow users to override the IP to use for the connection inside the
network namespace.

It is useful e.g. with slirp4netns to override the IP to "10.0.2.100".

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by giuseppe** [comment]
it is needed for Podman: https://github.com/containers/podman/issues/5138

**Comment by AkihiroSuda** [comment]
> it is needed for Podman: [containers/podman#5138](https://github.com/containers/podman/issues/5138)

This doesn't seem to be able to propagate src IP. Wouldn't this just make the src IP to 10.0.2.100 ?

**Comment by giuseppe** [comment]
Yes, it won't propagate the srcIP but at least we can set it to something different than 127.0.0.1 that causes issues with some container images

**Comment by giuseppe** [comment]
@AkihiroSuda are you fine with the change as it is?

File: pkg/port/builtin/child/child.go [hunk 1]
```diff
@@ -106,7 +106,11 @@ func (d *childDriver) handleConnectRequest(c *net.UnixConn, req *msg.Request) er
 		return errors.Errorf("unknown proto: %q", req.Proto)
 	}
 	var dialer net.Dialer
-	targetConn, err := dialer.Dial(req.Proto, fmt.Sprintf("127.0.0.1:%d", req.Port))
+	ip := req.IP
```
Comment (AkihiroSuda) [inline pkg/port/builtin/child/child.go:109]: We need to verify that this is a valid IPv4 string

File: pkg/port/builtin/child/child.go [hunk 2]
```diff
@@ -106,7 +106,11 @@ func (d *childDriver) handleConnectRequest(c *net.UnixConn, req *msg.Request) er
...
+	if ip == "" {
+		ip = "127.0.0.1"
+	}
+	targetConn, err := dialer.Dial(req.Proto, fmt.Sprintf("%s:%d", ip, req.Port))
```
Comment (AkihiroSuda) [inline pkg/port/builtin/child/child.go:113]: Could you also update pkg/port/{slirp4netns,socat}?

File: pkg/port/port.go [hunk 3]
```diff
@@ -10,6 +10,7 @@ type Spec struct {
 	ParentIP   string `json:"parentIP,omitempty"` // IPv4 address. can be empty (0.0.0.0).
 	ParentPort int    `json:"parentPort,omitempty"`
 	ChildPort  int    `json:"childPort,omitempty"`
+	ChildIP    string `json:"childIP,omitempty"`
```
Comment (AkihiroSuda) [inline pkg/port/port.go:13]: Please add godoc comment to clarify that this defaults to 127.0.0.1

**Review by AkihiroSuda — COMMENTED** [review]
Looks good but a couple of nits. Thanks

**Review by AkihiroSuda — APPROVED** [review]
Thanks, I'll ship a new release tomorrow

**Pull request merged**